File: apps/api/src/routes/logbook.ts
// SPDX-License-Identifier: Apache-2.0
import type { FastifyInstance } from 'fastify';
import type {
  LogbookProfile,
  LogbookExport,
  SessionScoreRecord,
  DimensionScore,
  PersonalRecord,
  ScenarioDifficulty,
} from '@convsim/shared';
import { getDb, WORKBENCH_TEST_SCENARIO_ID } from '../db.js';

interface SessionRow {
  session_id: string;
  scenario_id: string;
  ending_type: string | null;
  created_at: string;
  ended_at: string | null;
  setup_json: string;
  turn_count: number;
}

interface DebriefEventRow {
  session_id: string;
  payload_json: string;
  created_at: string;
}

// Returns YYYY-MM-DD from an ISO timestamp, using the timestamp as-is (UTC).
function toDateStr(iso: string): string {
  return iso.slice(0, 10);
}

// Recency-weighted average. scores[0] = most recent session score.
function computeRollingScore(scores: number[]): number {
  if (scores.length === 0) return 0;
  const DECAY = 0.85;
  let weightedSum = 0;
  let weightSum = 0;
  for (let i = 0; i < scores.length; i++) {
    const w = Math.pow(DECAY, i);
    weightedSum += scores[i] * w;
    weightSum += w;
  }
  return Math.round((weightedSum / weightSum) * 10) / 10;
}

// Count consecutive calendar days ending on today or yesterday.
function computeStreak(dates: string[]): number {
  if (dates.length === 0) return 0;

  const todayMs = Date.now();
  const todayStr = new Date(todayMs).toISOString().slice(0, 10);
  const yesterdayStr = new Date(todayMs - 86_400_000).toISOString().slice(0, 10);

  const unique = [...new Set(dates)].sort().reverse();
  if (unique[0] !== todayStr && unique[0] !== yesterdayStr) return 0;

  let streak = 1;
  let prev = unique[0];
  for (let i = 1; i < unique.length; i++) {
    const curr = unique[i];
    const prevMs = new Date(prev + 'T00:00:00Z').getTime();
    const currMs = new Date(curr + 'T00:00:00Z').getTime();
    if (Math.round((prevMs - currMs) / 86_400_000) === 1) {
      streak++;
      prev = curr;
    } else {
      break;
    }
  }
  return streak;
}

function buildProfile(db: ReturnType<typeof getDb>): LogbookProfile {
  const sessions = db
    .prepare<[string], SessionRow>(
      `SELECT session_id, scenario_id, ending_type, created_at, ended_at, setup_json, turn_count
       FROM sessions
       WHERE state = 'Ended' AND scenario_id != ?
       ORDER BY COALESCE(ended_at, created_at) DESC`,
    )
    .all(WORKBENCH_TEST_SCENARIO_ID);

  const totalSessions = sessions.length;

  let totalPracticeSeconds = 0;
  for (const s of sessions) {
    if (s.ended_at) {
      const diff = new Date(s.ended_at).getTime() - new Date(s.created_at).getTime();
      if (diff > 0) totalPracticeSeconds += Math.round(diff / 1000);
    }
  }

  const sessionDates = sessions.map((s) => toDateStr(s.ended_at ?? s.created_at));
  const streakDays = computeStreak(sessionDates);
  const lastSessionDate = sessions.length > 0 ? toDateStr(sessions[0].ended_at ?? sessions[0].created_at) : null;

  const debriefRows = db
    .prepare<[string], DebriefEventRow>(
      `SELECT se.session_id, se.payload_json, se.created_at
       FROM session_events se
       INNER JOIN sessions s ON s.session_id = se.session_id
       WHERE se.event_type = 'debrief_generated'
         AND s.scenario_id != ?
       ORDER BY se.event_id DESC`,
    )
    .all(WORKBENCH_TEST_SCENARIO_ID);

  // Index sessions by id for O(1) lookup
  const sessionById = new Map(sessions.map((s) => [s.session_id, s]));

  // dimension_id → scores ordered most-recent-first
  const dimScoresMap = new Map<string, number[]>();
  // (scenario_id|difficulty) → best personal record
  const prMap = new Map<string, { best_score: number; achieved_at: string; scenario_id: string; difficulty: string }>();
  // session_id → overall_score (first debrief per session wins)
  const overallBySession = new Map<string, number>();
  // Sessions already folded in. A session can be debriefed more than once
  // (e.g. the debrief screen's retry re-runs generation from the Ended state),
  // producing multiple debrief_generated events. Only the most recent debrief
  // per session should count — otherwise dimension session counts and the
  // recency-weighted rolling score are inflated by re-debriefs. debriefRows is
  // ordered event_id DESC, so the first row seen for a session is the newest.
  const seenSessions = new Set<string>();

  for (const row of debriefRows) {
    if (seenSessions.has(row.session_id)) continue;

    const payload = JSON.parse(row.payload_json) as {
      scores?: Record<string, number>;
      overall_score?: number;
    };

    const session = sessionById.get(row.session_id);
    if (!session) continue;
    seenSessions.add(row.session_id);

    const setup = JSON.parse(session.setup_json) as { difficulty: string };

    for (const [dimId, score] of Object.entries(payload.scores ?? {})) {
      if (!dimScoresMap.has(dimId)) dimScoresMap.set(dimId, []);
      dimScoresMap.get(dimId)!.push(score);
    }

    if (payload.overall_score !== undefined) {
      const prKey = `${session.scenario_id}|${setup.difficulty}`;
      const existing = prMap.get(prKey);
      if (!existing || payload.overall_score > existing.best_score) {
        prMap.set(prKey, {
          best_score: payload.overall_score,
          achieved_at: session.ended_at ?? session.created_at,
          scenario_id: session.scenario_id,
          difficulty: setup.difficulty,
        });
      }

      if (!overallBySession.has(row.session_id)) {
        overallBySession.set(row.session_id, payload.overall_score);
      }
    }
  }

  const dimensionScores: DimensionScore[] = Array.from(dimScoresMap.entries()).map(
    ([dimension_id, scores]) => ({
      dimension_id,
      rolling_score: computeRollingScore(scores),
      session_count: scores.length,
      // scores is most-recent-first; reverse to chronological for the chart.
      trajectory: [...scores].reverse(),
    }),
  );

  let strongestDimension: string | null = null;
  let weakestDimension: string | null = null;
  if (dimensionScores.length > 0) {
    const sorted = [...dimensionScores].sort((a, b) => b.rolling_score - a.rolling_score);
    strongestDimension = sorted[0].dimension_id;
    weakestDimension = sorted[sorted.length - 1].dimension_id;
  }

  // last_session_delta: last session's overall_score minus penultimate session's
  let lastSessionDelta: number | null = null;
  const scoresInOrder = sessions
    .map((s) => overallBySession.get(s.session_id))
    .filter((s): s is number => s !== undefined);
  if (scoresInOrder.length >= 2) {
    lastSessionDelta = scoresInOrder[0] - scoresInOrder[1];
  }

  const personalRecords: PersonalRecord[] = Array.from(prMap.values()).map((pr) => ({
    scenario_id: pr.scenario_id,
    difficulty: pr.difficulty as ScenarioDifficulty,
    best_score: pr.best_score,
    achieved_at: pr.achieved_at,
  }));

  return {
    total_sessions: totalSessions,
    total_practice_seconds: totalPracticeSeconds,
    streak_days: streakDays,
    last_session_date: lastSessionDate,
    dimension_scores: dimensionScores,
    personal_records: personalRecords,
    strongest_dimension: strongestDimension,
    weakest_dimension: weakestDimension,
    last_session_delta: lastSessionDelta,
  };
}

export async function logbookRoutes(app: FastifyInstance) {
  // GET /api/logbook/profile — aggregated training profile across all sessions
  app.get('/api/logbook/profile', async (): Promise<LogbookProfile> => {
    return buildProfile(getDb());
  });

  // GET /api/logbook/export — full JSON export (profile + per-session scores)
  app.get('/api/logbook/export', async (): Promise<LogbookExport> => {
    const db = getDb();
    const profile = buildProfile(db);

    const sessions = db
      .prepare<[string], SessionRow>(
        `SELECT session_id, scenario_id, ending_type, created_at, ended_at, setup_json, turn_count
         FROM sessions
         WHERE state = 'Ended' AND scenario_id != ?
         ORDER BY COALESCE(ended_at, created_at) DESC`,
      )
      .all(WORKBENCH_TEST_SCENARIO_ID);

    const debriefRows = db
      .prepare<[string], DebriefEventRow>(
        `SELECT se.session_id, se.payload_json
         FROM session_events se
         INNER JOIN sessions s ON s.session_id = se.session_id
         WHERE se.event_type = 'debrief_generated'
           AND s.scenario_id != ?
         ORDER BY se.event_id DESC`,
      )
      .all(WORKBENCH_TEST_SCENARIO_ID);

    const scoresBySession = new Map<string, { overall_score: number | null; scores: Record<string, number> }>();
    for (const row of debriefRows) {
      if (!scoresBySession.has(row.session_id)) {
        const payload = JSON.parse(row.payload_json) as {
          scores?: Record<string, number>;
          overall_score?: number;
        };
        scoresBySession.set(row.session_id, {
          overall_score: payload.overall_score ?? null,
          scores: payload.scores ?? {},
        });
      }
    }

    const sessionScores: SessionScoreRecord[] = sessions.map((s) => {
      const setup = JSON.parse(s.setup_json) as { difficulty: string };
      const debrief = scoresBySession.get(s.session_id);
      return {
        session_id: s.session_id,
        scenario_id: s.scenario_id,
        difficulty: setup.difficulty,
        ended_at: s.ended_at,
        overall_score: debrief?.overall_score ?? null,
        scores: debrief?.scores ?? {},
      };
    });

    return {
      exported_at: new Date().toISOString(),
      profile,
      session_scores: sessionScores,
    };
  });
}

File: services/convsim-core/convsim_core/routers/logbook.py
# SPDX-License-Identifier: Apache-2.0
"""Logbook endpoints: a local, aggregated skill profile across all sessions.

The Logbook is the player-facing record of practice — sessions played, hours
practiced, per-skill trajectory, consecutive-day streak, and personal records.
All data is derived from the local SQLite store (turn_sessions +
session_debriefs); nothing leaves the device and clearing local data clears the
logbook too (the /api/privacy/clear cascade removes the underlying rows).
"""
from __future__ import annotations

import json
from datetime import datetime, timedelta, timezone
from typing import Optional

from fastapi import APIRouter, Request
from pydantic import BaseModel

router = APIRouter(prefix="/api/logbook", tags=["logbook"])

# Temporary workbench test sessions are created under this scenario id and must
# never contribute to the player's training record (mirrors the /api/sessions
# exclusion so a lingering row cannot inflate the logbook).
_WORKBENCH_SCENARIO_ID = "workbench_test"

# Flow states that represent a completed practice session. A session is "Ended"
# once it terminates and "DebriefReady" once a debrief has been generated for
# it; both count as sessions played.
_COMPLETED_STATES = ("Ended", "DebriefReady")

# Exponential recency decay applied per-session when computing rolling skill
# scores: the most recent debriefed session carries weight 1.0, the previous
# 0.85, and so on. Recent form matters more than distant history.
_DECAY = 0.85


class DimensionScore(BaseModel):
    dimension_id: str
    rolling_score: float
    session_count: int
    # Per-session scores for this dimension in chronological order (oldest →
    # newest); raw material for the Logbook's per-skill trajectory chart.
    trajectory: list[float]


class PersonalRecord(BaseModel):
    scenario_id: str
    difficulty: str
    best_score: float
    achieved_at: str


class LogbookProfile(BaseModel):
    total_sessions: int
    total_practice_seconds: int
    streak_days: int
    last_session_date: Optional[str]
    dimension_scores: list[DimensionScore]
    personal_records: list[PersonalRecord]
    strongest_dimension: Optional[str]
    weakest_dimension: Optional[str]
    last_session_delta: Optional[float]


class SessionScoreRecord(BaseModel):
    session_id: str
    scenario_id: str
    difficulty: str
    ended_at: Optional[str]
    overall_score: Optional[float]
    scores: dict[str, float]


class LogbookExport(BaseModel):
    exported_at: str
    profile: LogbookProfile
    session_scores: list[SessionScoreRecord]


def _parse_dt(value: str) -> Optional[datetime]:
    """Parse a stored timestamp (SQLite 'YYYY-MM-DD HH:MM:SS' or ISO-8601).

    Stored timestamps are all UTC but inconsistently formatted: session
    created_at is written as tz-aware ISO ('...+00:00') while ended_at is
    written by SQLite's datetime('now') as a naive 'YYYY-MM-DD HH:MM:SS'.
    Normalise naive values to UTC so an aware created_at and a naive ended_at
    can be subtracted without raising "can't subtract offset-naive and
    offset-aware datetimes".
    """
    try:
        dt = datetime.fromisoformat(value)
    except (ValueError, TypeError):
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt


def _to_date_str(value: str) -> str:
    """Return YYYY-MM-DD from a stored timestamp, using it as-is (UTC)."""
    return value[:10]


def _rolling_score(scores: list[float]) -> float:
    """Recency-weighted average. scores[0] is the most recent session score."""
    if not scores:
        return 0.0
    weighted_sum = 0.0
    weight_sum = 0.0
    for i, score in enumerate(scores):
        w = _DECAY ** i
        weighted_sum += score * w
        weight_sum += w
    return round(weighted_sum / weight_sum, 1)


def _compute_streak(dates: list[str]) -> int:
    """Count consecutive calendar days (UTC) ending on today or yesterday."""
    if not dates:
        return 0

    today = datetime.now(timezone.utc).date()
    yesterday = today - timedelta(days=1)

    unique = sorted({d for d in dates}, reverse=True)
    first = datetime.fromisoformat(unique[0]).date()
    if first != today and first != yesterday:
        return 0

    streak = 1
    prev = first
    for d in unique[1:]:
        curr = datetime.fromisoformat(d).date()
        if (prev - curr).days == 1:
            streak += 1
            prev = curr
        else:
            break
    return streak


def _session_difficulty(setup_json: Optional[str]) -> str:
    try:
        return json.loads(setup_json or "{}").get("difficulty", "standard")
    except (ValueError, TypeError):
        return "standard"


def _build_profile(conn) -> LogbookProfile:
    placeholders = ",".join("?" for _ in _COMPLETED_STATES)
    sessions = conn.execute(
        f"""
        SELECT session_id, scenario_id, created_at, ended_at, setup_json
        FROM turn_sessions
        WHERE flow_state IN ({placeholders}) AND scenario_id != ?
        ORDER BY COALESCE(ended_at, created_at) DESC
        """,
        (*_COMPLETED_STATES, _WORKBENCH_SCENARIO_ID),
    ).fetchall()

    total_sessions = len(sessions)

    total_practice_seconds = 0
    for s in sessions:
        if s["ended_at"]:
            start = _parse_dt(s["created_at"])
            end = _parse_dt(s["ended_at"])
            if start and end:
                diff = (end - start).total_seconds()
                if diff > 0:
                    total_practice_seconds += round(diff)

    session_dates = [_to_date_str(s["ended_at"] or s["created_at"]) for s in sessions]
    streak_days = _compute_streak(session_dates)
    last_session_date = (
        _to_date_str(sessions[0]["ended_at"] or sessions[0]["created_at"])
        if sessions
        else None
    )

    # One debrief per session (session_debriefs has a unique index on
    # session_id), so no re-debrief de-duplication is needed here.
    debriefs = conn.execute(
        """
        SELECT sd.session_id, sd.content_json
        FROM session_debriefs sd
        INNER JOIN turn_sessions ts ON ts.session_id = sd.session_id
        WHERE ts.scenario_id != ?
        ORDER BY sd.id DESC
        """,
        (_WORKBENCH_SCENARIO_ID,),
    ).fetchall()

    session_meta = {
        s["session_id"]: {
            "scenario_id": s["scenario_id"],
            "difficulty": _session_difficulty(s["setup_json"]),
            "achieved_at": s["ended_at"] or s["created_at"],
        }
        for s in sessions
    }

    # dimension_id -> scores ordered most-recent-first
    dim_scores: dict[str, list[float]] = {}
    # (scenario_id|difficulty) -> best personal record
    pr_map: dict[str, PersonalRecord] = {}
    # session_id -> overall_score
    overall_by_session: dict[str, float] = {}

    for row in debriefs:
        meta = session_meta.get(row["session_id"])
        if meta is None:
            continue
        payload = json.loads(row["content_json"])

        for dim_id, score in (payload.get("scores") or {}).items():
            dim_scores.setdefault(dim_id, []).append(float(score))

        overall = payload.get("overall_score")
        if overall is not None:
            overall_by_session[row["session_id"]] = float(overall)
            pr_key = f"{meta['scenario_id']}|{meta['difficulty']}"
            existing = pr_map.get(pr_key)
            if existing is None or float(overall) > existing.best_score:
                pr_map[pr_key] = PersonalRecord(
                    scenario_id=meta["scenario_id"],
                    difficulty=meta["difficulty"],
                    best_score=float(overall),
                    achieved_at=meta["achieved_at"],
                )

    dimension_scores = [
        DimensionScore(
            dimension_id=dim_id,
            rolling_score=_rolling_score(scores),
            session_count=len(scores),
            # scores is most-recent-first; reverse to chronological for the chart.
            trajectory=list(reversed(scores)),
        )
        for dim_id, scores in dim_scores.items()
    ]

    strongest_dimension: Optional[str] = None
    weakest_dimension: Optional[str] = None
    if dimension_scores:
        ranked = sorted(dimension_scores, key=lambda d: d.rolling_score, reverse=True)
        strongest_dimension = ranked[0].dimension_id
        weakest_dimension = ranked[-1].dimension_id

    # last_session_delta: most recent session's overall score minus the previous
    # scored session's. sessions is ordered most-recent-first.
    last_session_delta: Optional[float] = None
    scores_in_order = [
        overall_by_session[s["session_id"]]
        for s in sessions
        if s["session_id"] in overall_by_session
    ]
    if len(scores_in_order) >= 2:
        last_session_delta = scores_in_order[0] - scores_in_order[1]

    return LogbookProfile(
        total_sessions=total_sessions,
        total_practice_seconds=total_practice_seconds,
        streak_days=streak_days,
        last_session_date=last_session_date,
        dimension_scores=dimension_scores,
        personal_records=list(pr_map.values()),
        strongest_dimension=strongest_dimension,
        weakest_dimension=weakest_dimension,
        last_session_delta=last_session_delta,
    )


@router.get("/profile", response_model=LogbookProfile)
async def get_logbook_profile(request: Request) -> LogbookProfile:
    """Aggregated training profile across all completed local sessions."""
    conn = request.app.state.db.connection()
    return _build_profile(conn)


@router.get("/export", response_model=LogbookExport)
async def export_logbook(request: Request) -> LogbookExport:
    """Full JSON export: the profile plus raw per-session score records."""
    conn = request.app.state.db.connection()
    profile = _build_profile(conn)

    placeholders = ",".join("?" for _ in _COMPLETED_STATES)
    sessions = conn.execute(
        f"""
        SELECT session_id, scenario_id, created_at, ended_at, setup_json
        FROM turn_sessions
        WHERE flow_state IN ({placeholders}) AND scenario_id != ?
        ORDER BY COALESCE(ended_at, created_at) DESC
        """,
        (*_COMPLETED_STATES, _WORKBENCH_SCENARIO_ID),
    ).fetchall()

    debriefs = conn.execute(
        """
        SELECT sd.session_id, sd.content_json
        FROM session_debriefs sd
        INNER JOIN turn_sessions ts ON ts.session_id = sd.session_id
        WHERE ts.scenario_id != ?
        ORDER BY sd.id DESC
        """,
        (_WORKBENCH_SCENARIO_ID,),
    ).fetchall()

    scores_by_session: dict[str, dict] = {}
    for row in debriefs:
        if row["session_id"] not in scores_by_session:
            payload = json.loads(row["content_json"])
            scores_by_session[row["session_id"]] = {
                "overall_score": payload.get("overall_score"),
                "scores": payload.get("scores") or {},
            }

    session_scores = []
    for s in sessions:
        debrief = scores_by_session.get(s["session_id"])
        session_scores.append(
            SessionScoreRecord(
                session_id=s["session_id"],
                scenario_id=s["scenario_id"],
                difficulty=_session_difficulty(s["setup_json"]),
                ended_at=s["ended_at"],
                overall_score=debrief["overall_score"] if debrief else None,
                scores=debrief["scores"] if debrief else {},
            )
        )

    return LogbookExport(
        exported_at=datetime.now(timezone.utc).isoformat(),
        profile=profile,
        session_scores=session_scores,
    )
